auth/src/main/java/software/amazon/awssdk/auth/signer/Aws4Signer.java

```diff
@@ -15,478 +15,99 @@
-            // request.
-            mutableRequest.rawQueryParameter(SignerConstants.X_AMZ_SECURITY_TOKEN,
-                                             ((AwsSessionCredentials) sanitizedCredentials).sessionToken());
-        }
-
-        final Aws4SignerRequestParams signerRequestParams = new Aws4SignerRequestParams(
-                execution.request(), mutableRequest, executionAttributes,
-                overriddenDate, regionName, serviceName, SignerConstants.AWS4_SIGNING_ALGORITHM);
-
-        // Add the important parameters for v4 signing
-        final String timeStamp = signerRequestParams.getFormattedSigningDateTime();
-
-        addPreSignInformationToRequest(mutableRequest, sanitizedCredentials, signerRequestParams, timeStamp, expirationInSeconds);
-
-        final String contentSha256 = calculateContentHashPresign(signerRequestParams, mutableRequest);
-
-        final String canonicalRequest = createCanonicalRequest(mutableRequest, contentSha256);
-
-        final String stringToSign = createStringToSign(canonicalRequest, signerRequestParams);
-
-        final byte[] signingKey = deriveSigningKey(sanitizedCredentials, signerRequestParams);
-
-        final byte[] signature = computeSignature(stringToSign, signingKey);
-
-        mutableRequest.rawQueryParameter(SignerConstants.X_AMZ_SIGNATURE, BinaryUtils.toHex(signature));
-
-        return mutableRequest.build();
-    }
-
-    /**
-     * Step 1 of the AWS Signature version 4 calculation. Refer to
-     * http://docs.aws
-     * .amazon.com/general/latest/gr/sigv4-create-canonical-request.html to
-     * generate the canonical request.
-     */
-    private String createCanonicalRequest(SdkHttpFullRequest.Builder request,
-                                          String contentSha256) {
-        final String canonicalRequest = request.method().toString() +
-                                        SignerConstants.LINE_SEPARATOR +
-                                        // This would optionally double url-encode the resource path
-                                        getCanonicalizedResourcePath(request.encodedPath(), doubleUrlEncode) +
-                                        SignerConstants.LINE_SEPARATOR +
-                                        getCanonicalizedQueryString(request.rawQueryParameters()) +
-                                        SignerConstants.LINE_SEPARATOR +
-                                        getCanonicalizedHeaderString(request.headers()) +
-                                        SignerConstants.LINE_SEPARATOR +
-                                        getSignedHeadersString(request.headers()) +
-                                        SignerConstants.LINE_SEPARATOR +
-                                        contentSha256;
-
-        if (LOG.isDebugEnabled()) {
-            LOG.debug("AWS4 Canonical Request: '\"" + canonicalRequest + "\"");
+        if (CredentialUtils.isAnonymous(signingParams.awsCredentials())) {
```

Comment:
I think with Zoe's changes it may be possible to get null credentials. We should handle that and throw if that's the case.

Reply:
Added validation when building Aws4SignerParams class